flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java

```diff
@@ -209,4 +211,15 @@ protected void handleStateChange(ConnectionState newState) {
 	public void unhandledError(String s, Throwable throwable) {
 		leaderListener.handleError(new FlinkException("Unhandled error in ZooKeeperLeaderRetrievalService:" + s, throwable));
 	}
+
+	private void notifyLeaderLoss() {
+		if (lastLeaderAddress != null || lastLeaderSessionID != null) {
+			LOG.debug(
+					"No leader information could be retrieved. Any listeners will be notified.");
+
+			lastLeaderAddress = null;
+			lastLeaderSessionID = null;
+			leaderListener.notifyLeaderAddress(null, null);
+		}
+	}
```

Comment:
This method is duplicating logic which is already contained in `ZooKeeperLeaderRetrievalService.nodeChanged`. I would suggest to factor out a method `notifyIfNewLeaderAddress(String newLeaderAddress, UUID newLeaderSessionID)`. `notifyLeaderLoss` could call this method with `notifyIfNewLeaderAddress(null, null)`.